packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-card.tsx

```diff
@@ -0,0 +1,244 @@
+  connectionOptions,
+  errors,
+  kmsProviderType,
+  fields,
+  clientCertIsOptional,
+  noTLS,
+  index,
+}: KMSProviderCardProps<T>) {
+  const [hoverProps, isHovered] = useHoverState();
+  const onRenameKmsProvider = useCallback(
+    (newName: KMSProviderName<T>) => {
+      return updateConnectionFormField({
+        type: 'rename-csfle-kms-provider',
+        name: kmsProviderName,
+        newName,
+      });
+    },
+    [kmsProviderName, updateConnectionFormField]
+  );
+  const onRemoveKmsProvider = useCallback(() => {
+    return updateConnectionFormField({
+      type: 'remove-csfle-kms-provider',
+      name: kmsProviderName,
+    });
+  }, [updateConnectionFormField, kmsProviderName]);
+
+  const onValidateName = useCallback(
+    (name: string) => {
+      // Exclude the current KMS provider name from the list.
+      const withoutCurrentKMSProviderNames = kmsProviderNames.filter(
+        (n) => n !== kmsProviderName
+      );
+      const maybeProviderName =
+        name === '' ? kmsProviderType : `${kmsProviderType}:${name}`;
+      // `kmsProviderType` can only exist once, so empty name is allowed only for that case.
+      if (
+        name === '' &&
+        withoutCurrentKMSProviderNames.some((n) => n === maybeProviderName)
+      ) {
+        return 'Name cannot be empty';
+      }
+      if (
+        withoutCurrentKMSProviderNames.includes(
+          maybeProviderName as KMSProviderName<T>
+        )
+      ) {
+        return 'Name already exists';
+      }
+      const regex = new RegExp(`^${kmsProviderType}(:[a-zA-Z0-9_]+)?$`);
+      if (!maybeProviderName.match(regex)) {
+        return 'Name must be alphanumeric and may contain underscores';
+      }
+      return undefined;
+    },
+    [kmsProviderNames, kmsProviderName, kmsProviderType]
+  );
+
+  return (
+    <Card
+      data-card-index={index}
+      data-testid={`${kmsProviderName}-kms-card-item`}
+      className={cardStyles}
+      {...hoverProps}
+    >
+      <div data-testid="kms-card-header" className={flexContainerStyles}>
+        <KMSNameComponent
+          kmsProviderName={kmsProviderName}
+          kmsProviderType={kmsProviderType}
+          onRename={onRenameKmsProvider}
+          validateName={onValidateName}
+        />
+        {kmsProviderNames.length > 1 && isHovered && (
+          <IconButton
+            aria-label="Remove KMS provider"
+            className={pushRightStyles}
+            onClick={onRemoveKmsProvider}
+          >
+            <Icon glyph="Trash" />
+          </IconButton>
+        )}
+      </div>
+      <KMSProviderFieldsForm
+        key={kmsProviderName}
+        errors={errors}
+        connectionOptions={connectionOptions}
+        updateConnectionFormField={updateConnectionFormField}
+        kmsProviderType={kmsProviderType}
+        kmsProviderName={kmsProviderName}
+        fields={fields}
+        clientCertIsOptional={clientCertIsOptional}
+        noTLS={noTLS}
+      />
+    </Card>
+  );
```

Comment:
Just wanted to say, I tried this out locally and it all feels very smooth and intuitive! This is amazing work ✨